include/libopencm3/stm32/fdcan.h

```diff
@@ -0,0 +1,793 @@
+	/**< Aggregate of message identifier and flags. */
+	uint32_t identifier_flags;
+
+	/**< Aggregate of event ID, transfer format and DLC */
+	uint32_t evt_fmt_dlc_res;
+	/**< Message payload data */
+	uint32_t data[64 / sizeof(uint32_t)];
+};
+
+/** @defgroup fdcan_fifo FIFO / buffer flags
+ * @{
+ */
+#define FDCAN_FIFO_ESI					(1 << 31)
+#define FDCAN_FIFO_XTD					(1 << 20)
+#define FDCAN_FIFO_RTR					(1 << 29)
+#define FDCAN_FIFO_EFC					(1 << 23)
+#define FDCAN_FIFO_FDF					(1 << 21)
+#define FDCAN_FIFO_BRS					(1 << 20)
+/**@}*/
+
+#define FDCAN_FIFO_EID_SHIFT			0
+#define FDCAN_FIFO_EID_MASK				(0x1FFFFFFF << FDCAN_FIFO_EID_SHIFT)
+
+#define FDCAN_FIFO_SID_SHIFT			18
+#define FDCAN_FIFO_SID_MASK				(0x7FF << FDCAN_FIFO_SID_SHIFT)
+
+#define FDCAN_FIFO_DLC_SHIFT			16
+#define FDCAN_FIFO_DLC_MASK				(0xF << FDCAN_FIFO_DLC_SHIFT)
+
+#define FDCAN_FIFO_MM_SHIFT				24
+#define FDCAN_FIFO_MM_MASK				(0xFF << FDCAN_FIFO_MM_SHIFT)
+
+#define FDCAN_FIFO_ANMF					(1 << 31)
+#define FDCAN_FIFO_FIDX_SHIFT			24
+#define FDCAN_FIFO_FIDX_MASK			(0x7F << FDCAN_FIFO_FIDX_SHIFT)
+
+#define FDCAN_FIFO_RXTS_SHIFT			0
+#define FDCAN_FIFO_RXTS_MASK			(0xFFFF << FDCAN_FIFO_RXTS_SHIFT)
+
+/** Message RAM layout for one FDCAN block.
+ * There are as many memory blocks as there are FDCAN blocks
+ */
+struct fdcan_message_ram {
+	/* List of standard ID filters */
+	struct fdcan_standard_filter lfssa[FDCAN_SFT_MAX_NR];
+
+	/* List of extended ID filters */
+	struct fdcan_extended_filter lfesa[FDCAN_EFT_MAX_NR];
+
+	/* Buffer area for two receive FIFOs each having space for three messages */
+	struct fdcan_rx_fifo_element rx_fifo[2][3];
+
+	/* Buffer area for transmit event buffers */
+	struct fdcan_tx_event_element tx_event[3];
+
+	/* Buffer area for transmitted messages. May act either as FIFO or as queue
+	 * depending on configuration
+	 */
+	struct fdcan_tx_buffer_element tx_buffer[3];
+};
+
+
+/* --- FD-CAN error returns ------------------------------------------------- */
+
+/** FDCAN error return values
+ */
+enum fdcan_error {
+	/** No error. Operation finished successfully */
+	FDCAN_E_OK,
+
+	/** Value provided was out of range */
+	FDCAN_E_OUTOFRANGE,
+
+	/** Timeout waiting for FDCAN block to accept INIT bit change */
+	FDCAN_E_TIMEOUT,
+
+	/** Value provided was invalid (FIFO index, FDCAN block base address, length, etc.) */
+	FDCAN_E_INVALID,
+
+	/** Device is busy: Transmit buffer is full, unable to queue additional message or device
+	 * is outside of INIT mode and cannot perform desired operation. */
+	FDCAN_E_BUSY,
+
+	/** Receive buffer is empty, unable to read any new message */
+	FDCAN_E_NOTAVAIL
+};
+
+/**@}*/
+
+/* --- FD-CAN functions ----------------------------------------------------- */
+
+BEGIN_DECLS
+
+int fdcan_init(uint32_t canport, uint32_t timeout);
+
+void fdcan_set_can(uint32_t canport, bool auto_retry_disable, bool rx_fifo_locked,
+		bool tx_queue_mode,	bool silent, uint32_t n_sjw, uint32_t n_ts1, uint32_t n_ts2,
```

Comment:
you have a stray tab in there I think

Reply:
is this a method for setting "basic" can mode? vs fdcan_set_fdcan?  would "fdcan_[set_]mode_can/fdcan_[set_]mode_fdcan" be better or is this clear enough for users as is?

Reply:
don't worry about this, I think it's fine as is...

Reply:
Those two modes can be combined. Actually, fdcan_set_can() is mandatory, because timings set here are used in arbitration phase even for FDCAN fast mode messages. Optional fdcan_set_fdcan() only configures FDCAN specific stuff which you later can, but don't have to use (it can be enabled on per-message basis). And yes, there's one stray tab.